# Marketplace Analytics Platform
## Notebook 1 — Raw Dataset Profiling & Structural Overview

This notebook establishes the structural facts of the Olist dataset before any analysis begins. It covers dataset shapes, memory usage, duplicate row counts, column data types, null counts, date field conversion, and overall date range coverage. The goal is orientation — understanding what data exists, how it is structured, and what time period it covers.

In [1]:
import sys
sys.path.append("../python")

from utils.imports import *
from utils.load_data import load_datasets

dfs = load_datasets()

In [2]:
pd.DataFrame([
    {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1]
    }
    for name, df in dfs.items()
])

,dataset,rows,columns
0,orders,99441,8
1,customers,99441,5
2,order_items,112650,7
3,payments,103886,5
4,reviews,99224,7
5,products,32951,9
6,sellers,3095,4
7,geolocation,1000163,5
8,category_translation,71,2


### Dataset Overview

Review the size and sample records from each dataset.

In [3]:
pd.DataFrame([
    {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
        "duplicate_rows": df.duplicated().sum(),
        "memory_mb": round(df.memory_usage(deep=True).sum() / 1024**2, 2)
    }
    for name, df in dfs.items()
])

,dataset,rows,columns,duplicate_rows,memory_mb
0,orders,99441,8,0,52.94
1,customers,99441,5,0,26.59
2,order_items,112650,7,0,35.99
3,payments,103886,5,0,16.23
4,reviews,99224,7,0,39.12
5,products,32951,9,0,6.30
6,sellers,3095,4,0,0.59
7,geolocation,1000163,5,261831,130.26
8,category_translation,71,2,0,0.01


In [4]:
for name, df in dfs.items():
    print("=" * 70)
    print(name.upper())
    display(df.head())

ORDERS


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


CUSTOMERS


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


ORDER_ITEMS


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


PAYMENTS


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


REVIEWS


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


PRODUCTS


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.00,287.00,1.00,225.00,16.00,10.00,14.00
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.00,276.00,1.00,"1,000.00",30.00,18.00,20.00
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.00,250.00,1.00,154.00,18.00,9.00,15.00
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.00,261.00,1.00,371.00,26.00,4.00,26.00
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.00,402.00,4.00,625.00,20.00,17.00,13.00


SELLERS


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


GEOLOCATION


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.55,-46.64,sao paulo,SP
1,1046,-23.55,-46.64,sao paulo,SP
2,1046,-23.55,-46.64,sao paulo,SP
3,1041,-23.54,-46.64,sao paulo,SP
4,1035,-23.54,-46.64,sao paulo,SP


CATEGORY_TRANSLATION


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


### Data Quality Assessment

Assess data types, completeness, duplicate rows, date fields.

In [5]:
for name, df in dfs.items():
    quality_view = pd.DataFrame({
        "column": df.columns,
        "data_type": df.dtypes.astype(str).values,
        "non_null_count": df.notna().sum().values,
        "null_count": df.isna().sum().values,
        "empty_string_count": [
            df[col].astype(str).str.strip().eq("").sum() if df[col].dtype == "object" else 0
            for col in df.columns
        ]
    })

    display(pd.DataFrame({"dataset": [name]}))
    display(quality_view)

,dataset
0,orders


,column,data_type,non_null_count,null_count,empty_string_count
0,order_id,object,99441,0,0
1,customer_id,object,99441,0,0
2,order_status,object,99441,0,0
3,order_purchase_timestamp,object,99441,0,0
4,order_approved_at,object,99281,160,0
5,order_delivered_carrier_date,object,97658,1783,0
6,order_delivered_customer_date,object,96476,2965,0
7,order_estimated_delivery_date,object,99441,0,0


,dataset
0,customers


,column,data_type,non_null_count,null_count,empty_string_count
0,customer_id,object,99441,0,0
1,customer_unique_id,object,99441,0,0
2,customer_zip_code_prefix,int64,99441,0,0
3,customer_city,object,99441,0,0
4,customer_state,object,99441,0,0


,dataset
0,order_items


,column,data_type,non_null_count,null_count,empty_string_count
0,order_id,object,112650,0,0
1,order_item_id,int64,112650,0,0
2,product_id,object,112650,0,0
3,seller_id,object,112650,0,0
4,shipping_limit_date,object,112650,0,0
5,price,float64,112650,0,0
6,freight_value,float64,112650,0,0


,dataset
0,payments


,column,data_type,non_null_count,null_count,empty_string_count
0,order_id,object,103886,0,0
1,payment_sequential,int64,103886,0,0
2,payment_type,object,103886,0,0
3,payment_installments,int64,103886,0,0
4,payment_value,float64,103886,0,0


,dataset
0,reviews


,column,data_type,non_null_count,null_count,empty_string_count
0,review_id,object,99224,0,0
1,order_id,object,99224,0,0
2,review_score,int64,99224,0,0
3,review_comment_title,object,11568,87656,2
4,review_comment_message,object,40977,58247,27
5,review_creation_date,object,99224,0,0
6,review_answer_timestamp,object,99224,0,0


,dataset
0,products


,column,data_type,non_null_count,null_count,empty_string_count
0,product_id,object,32951,0,0
1,product_category_name,object,32341,610,0
2,product_name_lenght,float64,32341,610,0
3,product_description_lenght,float64,32341,610,0
4,product_photos_qty,float64,32341,610,0
5,product_weight_g,float64,32949,2,0
6,product_length_cm,float64,32949,2,0
7,product_height_cm,float64,32949,2,0
8,product_width_cm,float64,32949,2,0


,dataset
0,sellers


,column,data_type,non_null_count,null_count,empty_string_count
0,seller_id,object,3095,0,0
1,seller_zip_code_prefix,int64,3095,0,0
2,seller_city,object,3095,0,0
3,seller_state,object,3095,0,0


,dataset
0,geolocation


,column,data_type,non_null_count,null_count,empty_string_count
0,geolocation_zip_code_prefix,int64,1000163,0,0
1,geolocation_lat,float64,1000163,0,0
2,geolocation_lng,float64,1000163,0,0
3,geolocation_city,object,1000163,0,0
4,geolocation_state,object,1000163,0,0


,dataset
0,category_translation


,column,data_type,non_null_count,null_count,empty_string_count
0,product_category_name,object,71,0,0
1,product_category_name_english,object,71,0,0


In [6]:
pd.DataFrame([
    {
        "dataset": name,
        "rows": df.shape[0],
        "duplicate_rows": df.duplicated().sum(),
        "duplicate_percentage": round(df.duplicated().mean() * 100, 2)
    }
    for name, df in dfs.items()
])

,dataset,rows,duplicate_rows,duplicate_percentage
0,orders,99441,0,0.00
1,customers,99441,0,0.00
2,order_items,112650,0,0.00
3,payments,103886,0,0.00
4,reviews,99224,0,0.00
5,products,32951,0,0.00
6,sellers,3095,0,0.00
7,geolocation,1000163,261831,26.18
8,category_translation,71,0,0.00


### Date Conversion

Date columns are initially loaded as text from CSV files. In this step, date fields are identified by name and converted to pandas datetime so date ranges and time-based patterns can be assessed correctly.

In [7]:
date_columns = {
    "orders": [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ],
    "order_items": ["shipping_limit_date"],
    "reviews": ["review_creation_date", "review_answer_timestamp"]
}

date_conversion_summary = []

for dataset, columns in date_columns.items():
    for column in columns:
        if column in dfs[dataset].columns:
            null_before = dfs[dataset][column].isna().sum()

            dfs[dataset][column] = pd.to_datetime(dfs[dataset][column], errors="coerce")

            null_after = dfs[dataset][column].isna().sum()
            coerced_to_null = null_after - null_before

            date_conversion_summary.append({
                "dataset": dataset,
                "column": column,
                "null_before_conversion": null_before,
                "coerced_to_null": coerced_to_null,
                "converted_count": dfs[dataset][column].notna().sum()
            })

date_conversion_summary_df = pd.DataFrame(date_conversion_summary)

date_conversion_summary_df

,dataset,column,null_before_conversion,coerced_to_null,converted_count
0,orders,order_purchase_timestamp,0,0,99441
1,orders,order_approved_at,160,0,99281
2,orders,order_delivered_carrier_date,1783,0,97658
3,orders,order_delivered_customer_date,2965,0,96476
4,orders,order_estimated_delivery_date,0,0,99441
5,order_items,shipping_limit_date,0,0,112650
6,reviews,review_creation_date,0,0,99224
7,reviews,review_answer_timestamp,0,0,99224


> **Note:** The date conversion logic above has been extracted into `python/utils/date_conversion.py` as a reusable utility function `convert_dates(dfs)`. All downstream notebooks import and call this function directly rather than repeating the conversion logic, ensuring date columns are consistently handled across the entire notebook series.

In [8]:
date_range_summary = []

all_min_dates = []
all_max_dates = []

for dataset, columns in date_columns.items():
    for column in columns:
        if column in dfs[dataset].columns:
            min_date = dfs[dataset][column].min()
            max_date = dfs[dataset][column].max()

            date_range_summary.append({
                "dataset": dataset,
                "column": column,
                "min_date": min_date.date() if pd.notna(min_date) else None,
                "max_date": max_date.date() if pd.notna(max_date) else None,
                "missing_values": dfs[dataset][column].isna().sum()
            })

            if pd.notna(min_date):
                all_min_dates.append(min_date)

            if pd.notna(max_date):
                all_max_dates.append(max_date)

date_range_summary_df = pd.DataFrame(date_range_summary)

display(date_range_summary_df)

overall_min_date = min(all_min_dates).date()
overall_max_date = max(all_max_dates).date()

print(f"Overall dataset date range: {overall_min_date} to {overall_max_date}")

,dataset,column,min_date,max_date,missing_values
0,orders,order_purchase_timestamp,2016-09-04,2018-10-17,0
1,orders,order_approved_at,2016-09-15,2018-09-03,160
2,orders,order_delivered_carrier_date,2016-10-08,2018-09-11,1783
3,orders,order_delivered_customer_date,2016-10-11,2018-10-17,2965
4,orders,order_estimated_delivery_date,2016-09-30,2018-11-12,0
5,order_items,shipping_limit_date,2016-09-19,2020-04-09,0
6,reviews,review_creation_date,2016-10-02,2018-08-31,0
7,reviews,review_answer_timestamp,2016-10-07,2018-10-29,0


Overall dataset date range: 2016-09-04 to 2020-04-09


> **⚑ Quality Flag — `shipping_limit_date` anomaly:** `shipping_limit_date` extends to 2020, approximately two years past the dataset's last order date (Oct 2018). Since this column represents a seller's handling deadline set at order time, a date this far out is unexpected. Investigated in `02_data_quality_check.ipynb` → *Investigating shipping_limit_date anomaly*.

## Notebook Summary

This notebook established the structural foundation of the Olist dataset prior to any analysis or quality validation.

| Dataset | Rows | Columns | Duplicate Rows | Total Nulls |
|---|---|---|---|---|
| orders | 99,441 | 8 | 0 | 4,447 |
| customers | 99,441 | 5 | 0 | 0 |
| order_items | 112,650 | 7 | 0 | 0 |
| payments | 103,886 | 5 | 0 | 0 |
| reviews | 99,224 | 7 | 0 | 146,188 |
| products | 32,951 | 9 | 0 | 4,460 |
| sellers | 3,095 | 4 | 0 | 0 |
| geolocation | 1,000,163 | 5 | 261,831 | 0 |
| category_translation | 71 | 2 | 0 | 0 |

**Key structural observations:**
- Dataset spans **2016-09-04 to 2018-10-17** — approximately 2 years of real Brazilian e-commerce transactions.
- `geolocation` is the only table with duplicate rows (261,831 / 26.2%) — expected behavior as the file logs multiple geocoded addresses per zip code prefix rather than one canonical row per prefix.
- `reviews` accounts for the majority of null values — driven by optional free-text fields (`review_comment_title`, `review_comment_message`) which customers are not required to fill.
- `orders` nulls are entirely in delivery/approval timestamp columns, explained by non-delivered order statuses.
- Two anomalies surfaced during date range assessment are investigated formally in `02_data_quality_check.ipynb`: `shipping_limit_date` extending to 2020 and geolocation coordinates falling outside Brazil's geographic bounding box.

<hr style="height:2px; background-color:black; border:none;">